src/ol/snap.js

```diff
@@ -0,0 +1,544 @@
+   * @type {ol.Map}
+   */
+  this.map = map;
+
+  /**
+   * The result of the snap.
+   * @type {ol.Coordinate}
+   * @api
+   */
+  this.snappedCoordinate = coordinate.slice();
+
+  /**
+   * The result of the snap.
+   * @type {boolean}
+   * @api
+   */
+  this.snapped = false;
+};
+goog.inherits(ol.SnapEvent, goog.events.Event);
+
+
+/**
+ * @enum {string}
+ */
+ol.SnapProperty = {
+  SNAPPING: 'snapping'
+};
+
+
+
+/**
+ * @classdesc
+ * Helper class for providing snap in ol.interaction.Pointer.
+ *
+ * Example:
+ *
+ *     var snap = new ol.Snap({
+ *       source: source,
+ *       interactions: [
+ *          modify, draw
+ *       ]
+ *     });
+ *
+ * @constructor
+ * @extends {ol.Object}
+ * @param {olx.SnapOptions=} opt_options Options.
+ * @api
+ */
+ol.Snap = function(opt_options) {
+
+  goog.base(this);
+
+  var options = goog.isDef(opt_options) ? opt_options : {};
+
+  var features;
+  if (goog.isDef(options.source)) {
+    features = options.source.getFeatures();
+  }
+  if (goog.isDef(options.features)) {
+    features = options.features;
+  }
+  if (goog.isArray(features)) {
+    features = new ol.Collection(features);
+  }
+  goog.asserts.assert(goog.isDef(features));
+
+  /**
+   * @type {Array.<ol.interaction.Pointer>}
+   * @private
+   */
+  this.interactions_ = options.interactions;
+  goog.asserts.assert(goog.isDef(this.interactions_));
+
+
+  /**
+   * @private
+   * @type {Object.<string, goog.events.Key>}
+   */
+  this.interactionSnapListenerKeys_ = {};
+
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.pixelTolerance_ = goog.isDef(options.pixelTolerance) ?
+      options.pixelTolerance : 10;
+
+
+  /**
+  * Segment RTree for each layer
+  * @type {Object.<*, ol.structs.RBush>}
+  * @private
+  */
+  this.rBush_ = new ol.structs.RBush();
+
+
+  /**
+  * @const
+  * @private
+  * @type {Object.<string, function(ol.Feature, ol.geom.Geometry)> }
+  */
+  this.SEGMENT_WRITERS_ = {
+    'Point': this.writePointGeometry_,
+    'LineString': this.writeLineStringGeometry_,
+    'LinearRing': this.writeLineStringGeometry_,
+    'Polygon': this.writePolygonGeometry_,
+    'MultiPoint': this.writeMultiPointGeometry_,
+    'MultiLineString': this.writeMultiLineStringGeometry_,
+    'MultiPolygon': this.writeMultiPolygonGeometry_,
+    'GeometryCollection': this.writeGeometryCollectionGeometry_
+  };
+
+
+  features.forEach(this.addFeature, this);
+  if (goog.isDef(options.source)) {
+    goog.events.listen(options.source, ol.source.VectorEventType.ADDFEATURE,
+        this.handleFeatureAdd_, false, this);
+    goog.events.listen(options.source, ol.source.VectorEventType.REMOVEFEATURE,
+        this.handleFeatureRemove_, false, this);
+  } else if (goog.isDef(options.features)) {
+    goog.events.listen(features, ol.CollectionEventType.ADD,
+        this.handleFeatureAdd_, false, this);
+    goog.events.listen(features, ol.CollectionEventType.REMOVE,
+        this.handleFeatureRemove_, false, this);
+  }
```

Comment:
Not really, because the Draw interaction can be configured with only a source (so there won't be a collection) or a collection (so there won't be a source).